Userland/DevTools/HackStudio/HackStudioWidget.cpp

```diff
@@ -907,10 +912,45 @@ void HackStudioWidget::set_current_editor_wrapper(RefPtr<EditorWrapper> editor_w
     update_tree_view();
 }
 
+void HackStudioWidget::rename_file(String const& old_name, String const& new_name)
+{
+    auto editor_or_none = m_all_editor_wrappers.first_matching([&old_name](auto const& editor) {
+        return editor->filename() == old_name;
+    });
+    if (editor_or_none.has_value()) {
+        (*editor_or_none)->set_filename(new_name);
+        (*editor_or_none)->set_name(new_name);
+    }
+
+    if (m_open_files.contains(old_name)) {
+        VERIFY(m_open_files_vector.remove_first_matching([&old_name](auto const& file) { return file == old_name; }));
+        m_open_files_vector.append(new_name);
+
+        ProjectFile* f = m_open_files.get(old_name).release_value();
+        m_open_files.set(new_name, *f);
+        m_open_files.remove(old_name);
+        m_open_files_view->model()->invalidate();
+    }
+
+    if (m_file_watcher->is_watching(old_name)) {
+        VERIFY(!m_file_watcher->remove_watch(old_name).is_error());
+        VERIFY(!m_file_watcher->add_watch(new_name, Core::FileWatcherEvent::Type::Deleted).is_error());
+    }
+}
+
 void HackStudioWidget::configure_project_tree_view()
 {
     m_project_tree_view->set_model(m_project->model());
     m_project_tree_view->set_selection_mode(GUI::AbstractView::SelectionMode::MultiSelection);
+    m_project_tree_view->set_editable(true);
+    m_project_tree_view->aid_create_editing_delegate = [](auto&) {
+        return make<GUI::StringModelEditingDelegate>();
+    };
+    m_project->model().on_rename_successful = [this](auto& absolute_old_path, auto& absolute_new_path) {
```

Comment:
Not sure if this is the best place where this callback should be set

Reply:
Yeah, this way this callback isn't set if we open another project after the HackStudioWidget has been initialized. I think `HackStudioWidget::open_project` is a good place for setting it.